### ISSUE TITLE
Move VertexColorMesh3D from scalismo-faces to scalismo

### ISSUE THREAD

**Issue by Ghazi-Bouabene** [unibas-gravis/scalismo#235]
This class is defined here: 
https://github.com/unibas-gravis/scalismo-faces/blob/master/src/main/scala/scalismo/faces/mesh/Mesh.scala#L43

The move would require to also move the RGBA class. The advantage would be that we could release [this change](https://github.com/marcelluethi/scalismo-ui-1/commit/22d15b7b13ad542188644ad63385e412c0e474dc) to the official scalismo-ui to visualize vertex colored meshes, without the dependency on scalismo-faces.

**Comment by BernhardEgger** [comment]
I'm not sure if this solves the problem in a sustainable way. Later we would perhaps like to also visualize a Textured Mesh, a Morphable Model (with color model) or a 2D PixelImage. Consequently those would then also go to scalismo? All of this would then drag along more and more stuff from faces needed to visualize this.

**Comment by Andreas-Forster** [comment]
I get your point, but the other options are to make scalismo-ui dependent on scalismo-faces or implement the same thing a second time. I am not sure if one of these is a better option? I also do not see that it is a real consequence that textured meshes and the morphable model with PixelImages will directly follow the vertex color mesh.

I see clearly the need for a vertex color mesh in scalismo-ui and see also the "fear" that moving it could mean scalismo soaks up scalismo-faces nearly as a whole.

**Comment by BernhardEgger** [comment]
It is not really a fear - I anyway almost always use both at the same time :)

What is exactly the problem of scalismo-ui beeing dependent on scalismo-faces?

**Comment by madsendennis** [comment]
Couldn't it quickly become a bit messy as Scalismo-UI and Scalismo-faces both depends on Scalismo. If Scalismo-UI also depends on Scalismo-faces it could end up with having to make a combined release of all 3? Or is this already done today?

**Comment by Ghazi-Bouabene** [comment]
The releases would be too "tied". Say we have a new concept in Scalismo (affecting only the medical area) and we would like to visualize it in the UI .. we would need to first roll out a new scalismo-faces version (even though there are no changes) in order to be able to release a new scalismo-ui version.

**Comment by Ghazi-Bouabene** [comment]
I agree with your argument Bernhard that this could gradually lead to wanting all of Scalismo faces into Scalismo in order to visualize more and more face-related things. 

The alternative would be to stick with the fact that scalismo-ui is a medical only visualization and have a different ui dedicated for face-related visualization. The problem there is that we would have to recode a lot of the internal tools we already have relying on scalismo-ui.  Could an **official** fork of scalismo-ui, depending on scalismo-faces be a solution?

**Comment by BernhardEgger** [comment]
I think both, your initially proposed solution or a separated scalismo-faces-ui would be feasible. With my initial concerns the second one could be slightly more sustainable. Would it make sense to build scalismo-faces-ui depending on scalismo-ui to avoid duplicated code?

**Comment by Andreas-Forster** [comment]
My problem is a bit the argument that we do not want to have to coordinate the releases. But if from the scalismo-faces side we want to change one of the real core things in scalismo-faces, the vertex color mesh, we would completely rely on scalismo after moving it to this project. 

This does not seem to be a good state for scalismo-faces developers and users.

**Comment by sschoenborn** [comment]
Even now, `scalismo-faces` actually already heavily depends on scalismo in many other core aspects (e.g. `TriangleMesh`, `SurfacePointProperty`). It would not be too surprising having color and the additional mesh class also in `scalismo`. Also, these core concepts in `scalismo-faces` do not change very frequently and are rather stable. In my opinion, this would be a good opportunity to discuss the future of the `scalismo-faces` project: what should still exist separately? How strongly should it depend on `scalismo`? (Didn't we even talk about a possible `scalismo-core` once?)

In the long run, I would argue against having two versions of `VertexColorMesh` (mainly because of the `RGBA` class, that will become really confusing), but it could still be a valid starting point for now - until everybody settles on a common version suitable for both libs.

**Comment by Andreas-Forster** [comment]
To wrap up an offline discussion:

- We decided to make a pull request as a discussion basis where we first move the VertexColorMesh and everything needed for a proof of concept into scalismo.
- Of course, based on the final version of the pull request it will then be decided if it should be merged into scalismo or not.
- In the above-mentioned discussion, we will also have a look at the impact of the pull request on scalismo-faces, which will lead to a decision if we can drop those classes from scalismo-faces and use the new ones from scalismo.
- Hopefully, we will end up merging the pull request and dropping the classes from scalismo-faces. So we could move in a subsequent pull request all other classes useful to the whole community (e.g. ColorNormalMesh & more) which were omitted in the first pull request.

May the discussion be continued.
